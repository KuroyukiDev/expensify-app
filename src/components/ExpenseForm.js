import React, {Component} from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';

const now = moment();
now.locale('ja');
console.log(now.format('YYYYå¹´ MMM Do (ddd)'));

class ExpenseForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      desc: props.expense ? props.expense.desc : '',
      note: props.expense ? props.expense.note : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: undefined
    };
  }
  
  onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({
        amount
      }));
    }
  };
  
  onDescChange = (e) => {
    const desc = e.target.value;
    this.setState(() => ({
      desc
    }));
  };
  
  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({
      note
    }));
  };
  
  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({
        createdAt
      }));
    }
  };
  
  onFocusChange = ({ focused: calendarFocused }) => {
    this.setState(() => ({
      calendarFocused
    }));
  };
  
  onEnterPress = (e) => {
    if(e.keyCode === 13 && e.shiftKey === false) {
      let note = e.target.value;
      this.setState(() => ({
        note
      }));
    } else if(e.keyCode === 13 && e.shiftKey === true) {
      this.onSubmit(e);
    }
  };
  
  onSubmit = (e) => {
    e.preventDefault();
    if (!this.state.desc || !this.state.amount) {
      // Set error state equal to:
      // 'Please provide description and amount'
      this.setState(() => ({
        error: 'Please provide description and amount.'
      }));
      
    } else {
      // Clear the error
      this.setState(() => ({
        error: undefined
      }));
      console.log('Submitted!');
      this.props.onFormSubmit({
        desc: this.state.desc,
        amount: parseFloat(this.state.amount) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };
  
  render() {
    return (
      <div>
        {
          this.state.error && (
            <h4 style={{color: 'red'}}>
              <em>
                { this.state.error }
              </em>
            </h4>
          )
        }
        
        <form onSubmit={this.onSubmit}>
          <div className="row">
            <div className="col-xs-8 form-group">
              <input
                className="form-control input-lg"
                type="text"
                placeholder="Description"
                autoFocus
                value={this.state.desc}
                onChange={this.onDescChange}
              />
            </div>
            <div className="col-xs-4 form-group">
              <input
                className="form-control input-lg"
                type="text"
                placeholder="Amount"
                value={this.state.amount}
                onChange={this.onAmountChange}
              />
            </div>
          </div>
          <div className="row">
            <div className="col-xs-1">
              
            </div>
            <div className="col-xs-6">
              <h4 className="text-left">
                Set Date:
              </h4>
              <SingleDatePicker
                date={this.state.createdAt}
                onDateChange={this.onDateChange}
                focused={this.state.calendarFocused}
                onFocusChange={this.onFocusChange}
                numberOfMonths={1}
                isOutsideRange={() => false}
                // orientation="vertical"
              />
              <br/>
            </div>
            <div className="col-xs-5">
              
            </div>
          </div>
          <div className="row">
            <div className="form-group">
              <textarea
                className="form-control input-lg"
                cols={35}
                rows={5}
                placeholder="Add a note for your expense (optional)"
                value={this.state.note}
                onKeyDown={this.onEnterPress}
                onChange={this.onNoteChange}
              />
            </div>
          </div>
          <br/>
          <br/>
          <button>
            Add Expense
          </button>
        </form>
        
        ExpenseForm
        
      </div>
    );
  }
}

export default ExpenseForm;