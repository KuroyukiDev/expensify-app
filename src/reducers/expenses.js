
// Expenses Reducer
import {
  ADD_EXPENSE,
  EDIT_EXPENSE,
  REMOVE_EXPENSE
} from "../shared/types";

const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case ADD_EXPENSE:
      return [
        ...state,
        action.expense
      ];
    case REMOVE_EXPENSE:
      return state.filter(({ id }) => {
        return id !== action.id;
      });
    case EDIT_EXPENSE:
      return state.map((expense) => {
        if (expense.id === action.id) {
          return {
            ...expense,
            ...action.updates
          };
        } else {
          return expense;
        }
      });
    default:
      return state;
  }
};
