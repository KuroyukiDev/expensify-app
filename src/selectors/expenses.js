import moment from 'moment';
// Get visible expenses

export default (expenses, {
  text, sortBy, sortMode, startDate, endDate
}) => {
  return expenses.filter((expense) => {
    const createdAtMoment = moment(expense.createdAt);
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
    const textMatch = expense.desc.toLowerCase().includes(text.toLowerCase());
    
    // Returns a sorted version of the expenses array
    return startDateMatch && endDateMatch && textMatch;
  }).sort((a, b) => {
    if (sortMode === 'hi2lo') {
      if (sortBy === 'date') {
        return a.createdAt < b.createdAt ? 1 : -1;
      } else {
        return a.amount < b.amount ? 1 : -1;
      }
    } else {
      if (sortBy === 'date') {
        return a.createdAt > b.createdAt ? 1 : -1;
      } else {
        return a.amount > b.amount ? 1 : -1;
      }
    }
  });
};
