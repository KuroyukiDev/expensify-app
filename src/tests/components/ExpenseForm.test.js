import React from 'react';
import { shallow } from 'enzyme';
import ExpenseForm from '../../components/ExpenseForm';
import expenses from '../../shared/testExpensesArr';

test('should render the ExpenseForm correctly', () => {
  const wrapper = shallow(<ExpenseForm/>);
  expect(wrapper).toMatchSnapshot();
});

test('should render the ExpenseForm correctly with a pre-loaded expense', () => {
  const wrapper = shallow(<ExpenseForm expense={expenses[0]}/>);
  expect(wrapper).toMatchSnapshot();
});
