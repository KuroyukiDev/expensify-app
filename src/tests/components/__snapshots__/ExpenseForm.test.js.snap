// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the ExpenseForm correctly 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <div
      className="row"
    >
      <div
        className="col-xs-8 form-group"
      >
        <input
          autoFocus={true}
          className="form-control input-lg"
          onChange={[Function]}
          placeholder="Description"
          type="text"
          value=""
        />
      </div>
      <div
        className="col-xs-4 form-group"
      >
        <input
          className="form-control input-lg"
          onChange={[Function]}
          placeholder="Amount"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-xs-1"
      />
      <div
        className="col-xs-6"
      >
        <h4
          className="text-left title-text"
        >
          Set Date:
        </h4>
        <SingleDatePicker
          anchorDirection="left"
          customCloseIcon={null}
          customInputIcon={null}
          date={"1970-01-01T00:00:00.000Z"}
          daySize={39}
          disabled={false}
          displayFormat={[Function]}
          enableOutsideDays={false}
          firstDayOfWeek={null}
          focused={false}
          hideKeyboardShortcutsPanel={false}
          horizontalMargin={0}
          id="date"
          initialVisibleMonth={null}
          inputIconPosition="before"
          isDayBlocked={[Function]}
          isDayHighlighted={[Function]}
          isOutsideRange={[Function]}
          isRTL={false}
          keepOpenOnDateSelect={false}
          monthFormat="MMMM YYYY"
          navNext={null}
          navPrev={null}
          numberOfMonths={1}
          onClose={[Function]}
          onDateChange={[Function]}
          onFocusChange={[Function]}
          onNextMonthClick={[Function]}
          onPrevMonthClick={[Function]}
          openDirection="down"
          orientation="horizontal"
          phrases={
            Object {
              "calendarLabel": "Calendar",
              "chooseAvailableDate": [Function],
              "clearDate": "Clear Date",
              "closeDatePicker": "Close",
              "dateIsUnavailable": [Function],
              "enterKey": "Enter key",
              "escape": "Escape key",
              "hideKeyboardShortcutsPanel": "Close the shortcuts panel",
              "homeEnd": "Home and end keys",
              "jumpToNextMonth": "Move forward to switch to the next month",
              "jumpToPrevMonth": "Move backward to switch to the previous month",
              "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
              select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
              "keyboardShortcuts": "Keyboard Shortcuts",
              "leftArrowRightArrow": "Right and left arrow keys",
              "moveFocusByOneDay": "Move backward (left) and forward (right) by one day",
              "moveFocusByOneMonth": "Switch months",
              "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week",
              "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week",
              "openThisPanel": "Open this panel",
              "pageUpPageDown": "page up and page down keys",
              "questionMark": "Question mark",
              "returnFocusToInput": "Return to the date input field",
              "selectFocusedDate": "Select the date in focus",
              "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel",
              "upArrowDownArrow": "up and down arrow keys",
            }
          }
          placeholder="Date"
          readOnly={false}
          renderCalendarInfo={null}
          renderDay={null}
          renderMonth={null}
          reopenPickerOnClearDate={false}
          required={false}
          screenReaderInputMessage=""
          showClearDate={false}
          showDefaultInputIcon={false}
          weekDayFormat="dd"
          withFullScreenPortal={false}
          withPortal={false}
        />
        <br />
      </div>
      <div
        className="col-xs-5"
      />
    </div>
    <div
      className="row"
    >
      <div
        className="form-group"
      >
        <textarea
          className="form-control input-lg"
          cols={35}
          onChange={[Function]}
          onKeyDown={[Function]}
          placeholder="Add a note for your expense (optional)"
          rows={5}
          value=""
        />
      </div>
    </div>
    <br />
    <button
      className="btn btn-primary"
    >
      Add Expense
    </button>
  </form>
  <br />
</div>
`;

exports[`should render the ExpenseForm correctly with a pre-loaded expense 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <div
      className="row"
    >
      <div
        className="col-xs-8 form-group"
      >
        <input
          autoFocus={true}
          className="form-control input-lg"
          onChange={[Function]}
          placeholder="Description"
          type="text"
          value="Gum"
        />
      </div>
      <div
        className="col-xs-4 form-group"
      >
        <input
          className="form-control input-lg"
          onChange={[Function]}
          placeholder="Amount"
          type="text"
          value="1.95"
        />
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-xs-1"
      />
      <div
        className="col-xs-6"
      >
        <h4
          className="text-left title-text"
        >
          Set Date:
        </h4>
        <SingleDatePicker
          anchorDirection="left"
          customCloseIcon={null}
          customInputIcon={null}
          date={"1970-01-01T00:00:00.000Z"}
          daySize={39}
          disabled={false}
          displayFormat={[Function]}
          enableOutsideDays={false}
          firstDayOfWeek={null}
          focused={false}
          hideKeyboardShortcutsPanel={false}
          horizontalMargin={0}
          id="date"
          initialVisibleMonth={null}
          inputIconPosition="before"
          isDayBlocked={[Function]}
          isDayHighlighted={[Function]}
          isOutsideRange={[Function]}
          isRTL={false}
          keepOpenOnDateSelect={false}
          monthFormat="MMMM YYYY"
          navNext={null}
          navPrev={null}
          numberOfMonths={1}
          onClose={[Function]}
          onDateChange={[Function]}
          onFocusChange={[Function]}
          onNextMonthClick={[Function]}
          onPrevMonthClick={[Function]}
          openDirection="down"
          orientation="horizontal"
          phrases={
            Object {
              "calendarLabel": "Calendar",
              "chooseAvailableDate": [Function],
              "clearDate": "Clear Date",
              "closeDatePicker": "Close",
              "dateIsUnavailable": [Function],
              "enterKey": "Enter key",
              "escape": "Escape key",
              "hideKeyboardShortcutsPanel": "Close the shortcuts panel",
              "homeEnd": "Home and end keys",
              "jumpToNextMonth": "Move forward to switch to the next month",
              "jumpToPrevMonth": "Move backward to switch to the previous month",
              "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
              select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
              "keyboardShortcuts": "Keyboard Shortcuts",
              "leftArrowRightArrow": "Right and left arrow keys",
              "moveFocusByOneDay": "Move backward (left) and forward (right) by one day",
              "moveFocusByOneMonth": "Switch months",
              "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week",
              "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week",
              "openThisPanel": "Open this panel",
              "pageUpPageDown": "page up and page down keys",
              "questionMark": "Question mark",
              "returnFocusToInput": "Return to the date input field",
              "selectFocusedDate": "Select the date in focus",
              "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel",
              "upArrowDownArrow": "up and down arrow keys",
            }
          }
          placeholder="Date"
          readOnly={false}
          renderCalendarInfo={null}
          renderDay={null}
          renderMonth={null}
          reopenPickerOnClearDate={false}
          required={false}
          screenReaderInputMessage=""
          showClearDate={false}
          showDefaultInputIcon={false}
          weekDayFormat="dd"
          withFullScreenPortal={false}
          withPortal={false}
        />
        <br />
      </div>
      <div
        className="col-xs-5"
      />
    </div>
    <div
      className="row"
    >
      <div
        className="form-group"
      >
        <textarea
          className="form-control input-lg"
          cols={35}
          onChange={[Function]}
          onKeyDown={[Function]}
          placeholder="Add a note for your expense (optional)"
          rows={5}
          value=""
        />
      </div>
    </div>
    <br />
    <button
      className="btn btn-primary"
    >
      Add Expense
    </button>
  </form>
  <br />
</div>
`;
